### PR TITLE
ci: Reenable ccache

### PR DESCRIPTION
**Pull request by ddn0**
Now that I actually know how actions/cache works, ccache should work as
expected.

clean build: 30 minutes
build time after restarting workflow: 2 minutes
Overall workflow time after restart: 15 minutes

### CODE REVIEW

**Comment by ddn0** [comment]
I've been noticing some timeouts from the newly enabled distributed tests. See https://github.com/KatanaGraph/katana/issues/141 for the tracking ticket. I may add a test repeat or something until someone can identify the root cause.

File: .github/workflows/cpp.yaml [hunk 1]
```diff
@@ -29,6 +29,27 @@ jobs:
...
+        #   2a. if there are multiple matching keys, return the most recent entry
+        #   3. Repeat from 1 for the default branch
+        #
+        # Thus, the ccache here behaves intuitively. If there was a previous
+        # entry on this branch use it. Otherwise, use the most recent entry
+        # from the default branch.
+        #
+        # Keys should be unique because once a cache entry is created it will
+        # never be updated by actions/cache.
+        key: ${{matrix.os}}-${{matrix.build_type}}-ccache-${{github.sha}}
```
Comment (arthurp) [inline .github/workflows/cpp.yaml:49]: `github.sha` is the full SHA of the current commit, not the branch name. So this key seems like it would be invalid every time a new commit is added. The restore-key below would trigger for the next commit, but that never having a direct match seems weird. Maybe you mean `github.ref`?

File: .github/workflows/cpp.yaml [hunk 2]
```diff
@@ -29,6 +29,27 @@ jobs:
...
+        #   2. key is a prefix of a restore key on the current branch
+        #   2a. if there are multiple matching keys, return the most recent entry
+        #   3. Repeat from 1 for the default branch
+        #
+        # Thus, the ccache here behaves intuitively. If there was a previous
+        # entry on this branch use it. Otherwise, use the most recent entry
+        # from the default branch.
+        #
+        # Keys should be unique because once a cache entry is created it will
+        # never be updated by actions/cache.
```
Comment (arthurp) [inline .github/workflows/cpp.yaml:48]: This documentation seem relevant to every ccache. Maybe move it to CONTRIBUTING.md since it's not specific to this one declaration.

File: lonestar/CMakeLists.txt [hunk 3]
```diff
@@ -163,7 +163,9 @@ if(GALOIS_ENABLE_DIST)
         DEPENDS setup-make${suffix}
         FIXTURES_REQUIRED setup${suffix}
         FIXTURES_SETUP ${suffix}
-        ENVIRONMENT GALOIS_DO_NOT_BIND_THREADS=1)
+        # OpenMPI triggers a lot of leak warnings. Disable it during
+        # sanitizer builds for now.
+        ENVIRONMENT "GALOIS_DO_NOT_BIND_THREADS=1;ASAN_OPTIONS=detect_leaks=0")
```
Comment (arthurp) [inline lonestar/CMakeLists.txt:168]: This doesn't seem to have anything to do with ccache. Maybe this should be in a separate PR.

File: .github/workflows/cpp.yaml [hunk 4]
```diff
@@ -169,14 +197,25 @@ jobs:
...
+          Debug)
+            BUILD_TYPE=Debug
+            NUM_TEST_THREADS=1
             SANITIZER=""
-          ;;
+            ;;
+          *)
+            echo Unknown build type: ${{matrix.build_type}}
+            exit 1
+            ;;
```
Comment (arthurp) [inline .github/workflows/cpp.yaml:218]: This and the other sanitizer changes below seem like they belong in a different PR.

File: .github/workflows/python.yaml [hunk 5]
```diff
@@ -28,6 +28,15 @@ jobs:
       with:
         key: ${{matrix.os}}-conan-packaging-${{hashFiles('config/conanfile.txt')}}
         path: ~/.conan/data
+    - name: Cache ccache objects
+      uses: actions/cache@v2
+      with:
+        # Python ccache cannot be reused from standard C++ ccache because
+        # Python builds run as root and will make unprivileged use difficult.
```
Comment (arthurp) [inline .github/workflows/python.yaml:35]: What does this mean? None of the ccaches are reused between different jobs.

Comment (ddn0) [reply .github/workflows/python.yaml:35]: From: https://github.com/actions/cache#cache-scopes

A cache could be reused between action workflows if it had a matching key. Thus, if the python builds shared the same naming convention as the cpp builds, they could share the same cache. Naming them differently avoids this.

File: .github/workflows/cpp.yaml (hunk 1 above)
Comment (ddn0) [reply .github/workflows/cpp.yaml:49]: If we did use `github.ref` then if someone force-pushed they'd always get the first cache and never any updated one.

Giving unique names lets the github cache be updated; though it does depend on multiple matches picking the newest match and creates situations where direct matches are not very common, but it does achieve the expected behavior that the cache contains the most recent work on a branch.

File: .github/workflows/cpp.yaml (hunk 4 above)
Comment (ddn0) [reply .github/workflows/cpp.yaml]: It's more a convenience for the reviewer since this is already a separate commit, but sure #152
